quickwit-storage/src/cache/disk/store.rs

```diff
@@ -0,0 +1,910 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::hash::{Hash, Hasher};
+use std::io::ErrorKind;
+use std::ops::Range;
+use std::os::unix::fs::FileExt;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::vec::IntoIter;
+use std::{cmp, io};
+
+use parking_lot::{Mutex, RwLock};
+
+use super::access_log::{open_access_log, AccessLogWriter, Metadata};
+use super::file::{FileEntry, FileKey};
+
+/// Opens a given directory as a file backed cache.
+///
+/// Note:
+///  This will purge any files not explicitly managed by the
+///  cache as part of a cleanup process on startup.
+pub(crate) fn open_disk_store(
+    base_path: &Path,
+    max_fd: usize,
+) -> io::Result<Store<FileBackedDirectory>> {
+    check_and_create_lock(base_path)?;
+
+    let log = open_access_log(base_path)?;
+
+    purge_unknown_files(base_path, &log.metadata)?;
+
+    let directory = FileBackedDirectory::new(base_path, max_fd, log.metadata);
+    let store = Store::new(directory, log.writer);
+
+    Ok(store)
+}
+
+/// Checks if the directory lock file already exists.
+///
+/// If the lock already exists an io error is returned.
+/// Otherwise a lockfile is created.
+///
+/// The implementation is in general, quite naive and doesn't
+/// attempt to re-acquire the lock if it fails the fist time.
+///
+/// TODO: Possibly improve this functionality.
+fn check_and_create_lock(base_path: &Path) -> io::Result<()> {
+    let lock_file = base_path.join(".cache-lock");
+    if lock_file.exists() {
+        return Err(io::Error::new(
+            ErrorKind::Other,
+            "Existing lock file detected, directory already in use!",
+        ));
+    }
+
+    File::options().create(true).open(lock_file)?;
+
+    Ok(())
+}
+
+/// Attempts to delete the directory lock file.
+fn try_remove_lock(base_path: &Path) -> io::Result<()> {
+    let lock_file = base_path.join(".cache-lock");
+    std::fs::remove_file(lock_file)
+}
```

Comment:
The directory lock system is very naive and could really do with being improved, even if it just adds something like an exponential backoff for retrying to acquire the lock.